rust/kernel/thread.rs

```diff
@@ -0,0 +1,404 @@
+            }
+        }
+    };
+}
+primitive_thread_arg! {i8}
+primitive_thread_arg! {i16}
+primitive_thread_arg! {i32}
+primitive_thread_arg! {i64}
+primitive_thread_arg! {isize}
+primitive_thread_arg! {u8}
+primitive_thread_arg! {u16}
+primitive_thread_arg! {u32}
+primitive_thread_arg! {u64}
+primitive_thread_arg! {usize}
+
+/// Bridge function from Rust ABI to C.
+extern "C" fn bridge<A, F: ThreadFunctor<A>>(data: *mut c_types::c_void) -> i32
+where
+    A: ThreadArg,
+{
+    // SAFETY: `data` is the result of `A::into_pointer`, therefore it's safe
+    // to `from_pointer` here.
+    let arg = unsafe { A::from_pointer(data) };
+
+    match F::func(arg) {
+        Ok(_) => 0,
+        Err(e) => e.to_kernel_errno(),
+    }
+}
+
+/// Helper struct for thread that just call its argument as a closure.
+///
+/// We use double boxing in the implementation, because 1) we need to make sure
+/// the thread argument live longer enough for the new thread to use (box #1),
+/// and 2) `dyn FnOnce()` is a fat pointer that doesn't fit in `*mut c_void`
+/// (box #2).
+struct ClosureCallFunctor;
+
+impl ThreadFunctor<Box<Box<dyn FnOnce() -> KernelResult + Send>>> for ClosureCallFunctor {
+    fn func(arg: Box<Box<dyn FnOnce() -> KernelResult + Send>>) -> KernelResult {
+        // Just calls it.
+        arg()
+    }
+}
+
+/// A kernel thread handle.
+pub struct Thread {
+    /// Pointer to the kernel thread.
+    task: *mut bindings::task_struct,
+}
+
+impl Thread {
+    /// Creates a new thread using a C-style function pointer.
+    ///
+    /// No extra dynamic memory allocation for thread creation. Use this when
+    /// closure allocation overhead is unacceptable or there is already a C
+    /// style thread function.
+    ///
+    /// # Safety
+    ///
+    /// This function actually doesn't dereference `arg` or call `f`, so even if
+    /// the users pass incorrect parameters this function won't run into
+    /// trouble. But if the users provide incorrect `arg` or `f` the new
+    /// thread will corrupt memory or do other unsafe behaviors, so
+    /// make it `unsafe`. Otherwise, it is [`Thread::wake_up`] that should be
+    /// made as `unsafe` because it will trigger the call of the unsafe function
+    /// `f`, this is undesirable given that [`Thread::wake_up`] on a
+    /// [`Thread::try_new`] created thread is safe.
+    ///
+    /// The safety requirements of calling this function are:
+    ///
+    /// - Make sure `arg` is a proper pointer that points to a valid memory
+    ///   location when the new thread begins to run.
+    ///
+    /// - Make sure `f` is a proper function pointer and `f` handles `arg`
+    ///   correctly.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    pub unsafe fn try_new_extern(
+        name: CStr,
```

Comment:
Is this lifetime correct, or does `name` need to outlive this call? I'm not sure if `kthread_create_on_node` copies it, or needs this io live for hte lifetime of the thread.

Reply:
`kthread_create_on_node` will copy the content of the name as the SAFETY comments below says.